src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc

```diff
@@ -16,41 +16,261 @@
+  }
+  s2a_ServerSessionStartReq_set_in_bytes(
+      start_server, upb_strview_make(reinterpret_cast<const char*>(
+                                         GRPC_SLICE_START_PTR(*bytes_received)),
+                                     GRPC_SLICE_LENGTH(*bytes_received)));
+  return s2a_get_serialized_session_req(request, arena.ptr());
 }
 
-tsi_result s2a_handshaker_client_start_server(
-    const s2a_handshaker_client* client, grpc_slice* bytes_received) {
-  return TSI_UNIMPLEMENTED;
+tsi_result s2a_handshaker_client::server_start(grpc_slice* bytes_received) {
+  GPR_ASSERT(bytes_received != nullptr);
+  grpc_byte_buffer* buffer = s2a_get_serialized_start_server(bytes_received);
+  if (buffer == nullptr) {
+    gpr_log(GPR_ERROR, kS2AGetSerializedStartServerFailed);
+    return TSI_INTERNAL_ERROR;
+  }
+  grpc_byte_buffer_destroy(send_buffer_);
+  send_buffer_ = buffer;
+  tsi_result call_result = make_grpc_call(/*is_start=*/true);
+  if (call_result != TSI_OK) {
+    gpr_log(GPR_ERROR, kS2AMakeGrpcCallFailed);
+  }
+  return call_result;
 }
 
-tsi_result s2a_handshaker_client_next(const s2a_handshaker_client* client,
-                                      grpc_slice* bytes_received) {
-  return TSI_UNIMPLEMENTED;
+/** ------------ Preparation of next messages. --------------- **/
+
+grpc_byte_buffer* s2a_handshaker_client::s2a_get_serialized_next(
+    grpc_slice* bytes_received) {
+  GPR_ASSERT(bytes_received != nullptr);
+  upb::Arena arena;
+  s2a_SessionReq* request = s2a_SessionReq_new(arena.ptr());
+  s2a_SessionNextReq* next = s2a_SessionReq_mutable_next(request, arena.ptr());
+  s2a_SessionNextReq_set_in_bytes(
+      next, upb_strview_make(reinterpret_cast<const char*>(
+                                 GRPC_SLICE_START_PTR(*bytes_received)),
+                             GRPC_SLICE_LENGTH(*bytes_received)));
+  return s2a_get_serialized_session_req(request, arena.ptr());
 }
 
-void s2a_handshaker_client_shutdown(const s2a_handshaker_client* client) {
+tsi_result s2a_handshaker_client::next(grpc_slice* bytes_received) {
+  GPR_ASSERT(bytes_received != nullptr);
+  grpc_slice_unref_internal(recv_bytes_);
+  recv_bytes_ = grpc_slice_ref_internal(*bytes_received);
+  grpc_byte_buffer* buffer = s2a_get_serialized_next(bytes_received);
+  if (buffer == nullptr) {
+    gpr_log(GPR_ERROR, kS2AGetSerializedNextFailed);
+    return TSI_INTERNAL_ERROR;
+  }
+  grpc_byte_buffer_destroy(send_buffer_);
+  send_buffer_ = buffer;
+  tsi_result call_result = make_grpc_call(/*is_start=*/false);
+  if (call_result != TSI_OK) {
+    gpr_log(GPR_ERROR, kS2AMakeGrpcCallFailed);
+  }
+  return call_result;
+}
+
+/** ------------------- Callback methods. -------------------------------- **/
+
+static void s2a_on_status_received(void* arg, grpc_error* error) {
+  // TODO(mattstev): implement.
   return;
 }
 
+/** ------------------- Create, shutdown, & destroy methods. ------------- **/
+
+s2a_handshaker_client::s2a_handshaker_client(
```

Comment:
Can we use C++ initialization list for the constructor?

Reply:
Done in next commit.